### PR TITLE
PP-1098: PTL should cleanup job folders

### PR DESCRIPTION
**Pull request by kjakkali**
#### Issue-ID

* **[PP-1098](https://pbspro.atlassian.net/browse/PP-1098)**

#### Problem description
* PTL does not cleanup left over job dir under TEMP (ex : /var/tmp) at the end of pbs_benchpress execution.

#### Cause / Analysis
* The existing cleanup() just cleans up dejagnu temp dir , PTL temp files .

#### Solution description
* Update clenaup() to cleanup job dir also

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/fw/ptl/utils/plugins/ptl_test_runner.py [hunk 1]
```diff
@@ -635,22 +635,36 @@ def addSuccess(self, test):
...
+        for tmpdir in dirlist:
+            self.logger.info('Cleaning up ' + tmpdir + ' dir')
+            ftd = []
             files = du.listdir(path=tmpdir)
             bn = os.path.basename
             ftd.extend([f for f in files if bn(f).startswith('PtlPbs')])
             ftd.extend([f for f in files if bn(f).startswith('STDIN')])
+            ftd.extend([f for f in files if bn(f).startswith('pbsscrpt')])
+            ftd.extend([f for f in files if bn(f).startswith('pbs.conf.')])
+            ftd.extend([f for f in files if bn(f).startswith('pbs.')])
```
Comment (anamikau) [inline test/fw/ptl/utils/plugins/ptl_test_runner.py]: I suggest updating it with a regular expression to check pbs.\d+

File: test/fw/ptl/utils/plugins/ptl_test_runner.py [hunk 2]
```diff
@@ -635,22 +635,36 @@ def addSuccess(self, test):
...
-        ftd = []
-        if tmpdir:
+        dirlist = set(['/tmp', '/var/tmp'])
+        # get tmp dir from the environment
+        for envname in 'TMPDIR', 'TEMP', 'TMP':
+            dirname = os.getenv(envname)
+            if dirname: 
+                dirlist.add(dirname)
+
+        for tmpdir in dirlist:
```
Comment (anamikau) [inline test/fw/ptl/utils/plugins/ptl_test_runner.py]: Add comment of what you are trying to do here.

File: test/fw/ptl/utils/plugins/ptl_test_runner.py [hunk 3]
```diff
@@ -635,22 +635,36 @@ def addSuccess(self, test):
...
             bn = os.path.basename
             ftd.extend([f for f in files if bn(f).startswith('PtlPbs')])
             ftd.extend([f for f in files if bn(f).startswith('STDIN')])
+            ftd.extend([f for f in files if bn(f).startswith('pbsscrpt')])
+            ftd.extend([f for f in files if bn(f).startswith('pbs.conf.')])
+            ftd.extend([f for f in files if bn(f).startswith('pbs.')])
             for f in ftd:
                 du.rm(path=f, sudo=True, recursive=True, force=True,
                       level=logging.DEBUG)
+        #Dejagnu testsuite creates temp dir under tempfile.gettempdir()
```
Comment (anamikau) [inline test/fw/ptl/utils/plugins/ptl_test_runner.py]: remove the reference for Dejagnu

**Comment by bhroam** [comment]
If PBS itself is leaving temp files/directories around, then PBS has a bug.  Fixing PTL will just cover up the issue.  PBS should be cleaning up its temporary files/directories.

File: test/fw/ptl/utils/plugins/ptl_test_runner.py (hunk 1 above)
Comment (kjakkali) [reply test/fw/ptl/utils/plugins/ptl_test_runner.py]: @anamikau : 'pbs.' will serve the purpose. Do you still prefer regex over string match?

File: test/fw/ptl/utils/plugins/ptl_test_runner.py (hunk 2 above)
Comment (kjakkali) [reply test/fw/ptl/utils/plugins/ptl_test_runner.py]: done

File: test/fw/ptl/utils/plugins/ptl_test_runner.py (hunk 3 above)
Comment (kjakkali) [reply test/fw/ptl/utils/plugins/ptl_test_runner.py]: done

**Comment by kjakkali** [comment]
@bhroam : This issue is with different instances of PBS. Let me explain,

 This issue occurred in our in house automation tool. We run PTL test as follows 
 - install PBS -> install PTL -> run Test -> uninstall PBS. 
Each time pbs_benchpress command invoked with fresh installation of PBS. 
If previous test execution failed and uninstalled PBS then in next execution new instance of PBS does not recognize the job dir of old instance of PBS. 

So this fix is to clean up left over job dir before exiting pbs_benchpress command.

HTH.

Let me know you comments/suggestion.

File: test/fw/ptl/utils/plugins/ptl_test_runner.py (hunk 1 above)
Comment (anamikau) [reply test/fw/ptl/utils/plugins/ptl_test_runner.py]: @kjakkali the reason I asked you to check that there is a digit to make sure you are deleting jobdir only. There might be other files that one might have created with prefix "pbs." and we should not just be cleaning everything that starts with 'pbs.'.
Do you agree?

Comment (kjakkali) [reply test/fw/ptl/utils/plugins/ptl_test_runner.py]: done

File: test/fw/ptl/utils/plugins/ptl_test_runner.py [hunk 4]
```diff
@@ -642,6 +642,7 @@ def _cleanup(self):
             if dirname:
                 dirlist.add(dirname)
 
+        p = re.compile(r'^pbs\.\d+')
         for tmpdir in dirlist:
             #list the contents of each tmp dir and
             #get the file list to be deleted
```
Comment (bhroam) [inline test/fw/ptl/utils/plugins/ptl_test_runner.py]: I think pep8 will get mad at you for not having spaces after the #

**Comment by agrawalravi90** [comment]
"If previous test execution failed and uninstalled PBS then in next execution new instance of PBS does not recognize the job dir of old instance of PBS."

@kjakkali , but temporary job directories are being created by PBSPro itself (and not PTL), right? Shouldn't PBSPro itself cleanup whatever temporary directories it creates? I think that If there is something that PTL itself creates only then we should fix it in PTL.

**Comment by anamikau** [comment]
@agrawalravi90 I have filed following issues for PBS and PTL respectively.

PP-1203 : logged this to fix the PBS issue
PP1204: logged this for PTL enhancement to catch instance of PP-1203.

This issue is for PTL to start the new test fresh and make sure there are no left over files from previous installation that might cause failures.

**Comment by agrawalravi90** [comment]
Thanks for explaining Anamika, I'm ok with PTL taking care that the new installation gets a clean system.

**Comment by kjakkali** [comment]
@anamikau  : Codacy is complaining about hard coding temp dir path. It suggested to use tempfile module. tempfile module has function (tempfile.gettempdir())to get the system configured temp dir. But in my case we need to clean up all temp dirs . So we need to ignore codacy warning in this case.   

@bhroam : Please let me know if you have any comments.

**Comment by mike0042** [comment]
Could it be the case that multiple tests are running on the same system at the same time? If so, would removing all PTL temporary files affect the other running tests?

**Comment by bhroam** [comment]
The only thing I might suggest is to use os.path.join() instead of using /tmp and /var/tmp.  Since the two paths are only for linux, it won't help if python created a windows path out of it.  It's up to you.  I'll sign off as is.

**Comment by kjakkali** [comment]
@mike0042  : In jenkin we build and test only one build at a time.

**Comment by mike0042** [comment]
Thank you for confirming @kjakkali.

**Comment by mike0042** [comment]
Please squash and rebase for merge.

**Comment by kjakkali** [comment]
Squashed & rebased. Ready for merge.
@PBSPro/pbspro-maintainers  : Please take this PR forward.

**Comment by bhroam** [comment]
Sorry about this, but something went in before you.  Could you rebase again?

**Comment by kjakkali** [comment]
rebased again. 
@PBSPro/pbspro-maintainers : Please take this PR forward.

**Comment by bhroam** [comment]
@kjakkali could you post something in the ticket or the PR saying this doesn't require a PTL test.  I merged in the PR anyway.

**Pull request merged**